File: app/src/main/java/com/mkrworld/libfilter/dto/FilterMatrix.java
package com.mkrworld.libfilter.dto;

import com.mkrworld.libfilter.enums.FilterCategory;

public class FilterMatrix {

    private float[] mMatrix;
    private float mMultiplier;
    private FilterCategory mFilterCategory;
    private Offset mOffSet;

    /**
     * Constructor
     *
     * @param filterCategory
     * @param matrix         3X3 mMatrix
     */
    public FilterMatrix(FilterCategory filterCategory, float[] matrix) {
        this(filterCategory, new Offset(0F), 1F, matrix);
    }

    /**
     * Constructor
     *
     * @param filterCategory
     * @param multiplier
     * @param matrix         3X3 mMatrix
     */
    public FilterMatrix(FilterCategory filterCategory, Float multiplier, float[] matrix) {
        this(filterCategory, new Offset(0F), multiplier, matrix);
    }

    /**
     * Constructor
     *
     * @param filterCategory
     * @param offSet
     * @param matrix         3X3 mMatrix
     */
    public FilterMatrix(FilterCategory filterCategory, Offset offSet, float[] matrix) {
        this(filterCategory, offSet, 1F, matrix);
    }

    /**
     * Constructor
     *
     * @param filterCategory
     * @param offSet
     * @param multiplier
     * @param matrix         3X3 mMatrix
     */
    public FilterMatrix(FilterCategory filterCategory, Offset offSet, Float multiplier, float[] matrix) {
        this.mFilterCategory = filterCategory;
        this.mMultiplier = multiplier;
        this.mOffSet = offSet;
        this.mMatrix = getMatrix(filterCategory, offSet, matrix);
    }

    /**
     * Method to create Filter mMatrix from Input mMatrix
     */
    private float[] getMatrix(FilterCategory filterCategory, Offset offSet, float[] matrix) {
        if (matrix.length == 0) {
            switch (filterCategory) {
                case COLOR:
                    matrix = new float[]{
                            1F, 0F, 0F,
                            0F, 1F, 0F,
                            0F, 0F, 1F};
                    break;
                case CONVENTIONAL:
                    matrix = new float[]{
                            0F, 0F, 0F,
                            0F, 1F, 0F,
                            0F, 0F, 10F};
                    break;
                default:
                    matrix = new float[]{0F};
            }
        }
        switch (filterCategory) {
            case COLOR:
                return new float[]{
                        matrix[0], matrix[1], matrix[2], 0F, offSet.getmR(),
                        matrix[3], matrix[4], matrix[5], 0F, offSet.getmG(),
                        matrix[6], matrix[7], matrix[8], 0F, offSet.getmB(),
                        0F, 0F, 0F, 1F, 0F};
            case CONVENTIONAL:
                return new float[]{
                        matrix[0], matrix[1], matrix[2],
                        matrix[3], matrix[4], matrix[5],
                        matrix[6], matrix[7], matrix[8]};
            default:
                return new float[]{0F};
        }
    }

    /**
     * Method to get the Filter Category
     *
     * @return
     */
    public FilterCategory getFilterCategory() {
        return mFilterCategory;
    }

    /**
     * Method to get the Filter Multiplier
     *
     * @return
     */
    public float getMultiplier() {
        return mMultiplier;
    }

    /**
     * Method to get the Filter Matrix.<br>
     * <UL>
     * <li>4X5 for Color</li>
     * <li>3X3 for Conventional</li>
     * <li>1X1 for Rest Other</li>
     * </UL>
     *
     * @return
     */
    public float[] getMatrix() {
        return mMatrix;
    }

    /**
     * Method to get the OffSet
     *
     * @return
     */
    public Offset getOffSet() {
        return mOffSet;
    }
}

File: app/src/main/java/com/mkrworld/libfilter/dto/FilterMatrix.kt
package com.mkrworld.libfilter.dto

import com.mkrworld.libfilter.enums.FilterCategory

/**
 * @author THEMKR
 */
class FilterMatrix {

    /**
     * Method to get the Filter Matrix.<br></br>[<UL><LI>4X5 for Color</LI><LI>3X3 for Conventional</LI><LI>1X1 for Rest Other</LI></UL>]
     *
     * @return
     */
    val matrix: FloatArray

    /**
     * Method to get the Filter Multiplier
     *
     * @return
     */
    val multiplier: Float

    /**
     * Method to get the Filter Category
     *
     * @return
     */
    val filterCategory: FilterCategory

    /**
     * Method to get the OffSet
     *
     * @return
     */
    val offSet: Offset

    /**
     * Constructor
     *
     * @param filterCategory Category of effect
     * @param matrix [<UL><LI>4X5 for Color</LI><LI>3X3 for Conventional</LI><LI>1X1 for Rest Other</LI></UL>]
     */
    constructor(filterCategory: FilterCategory, matrix: FloatArray) : this(filterCategory, Offset(0f), 1f, matrix) {}

    /**
     * Constructor
     *
     * @param filterCategory Category of effect
     * @param multiplier Effect Value Multiplier
     * @param matrix [<UL><LI>4X5 for Color</LI><LI>3X3 for Conventional</LI><LI>1X1 for Rest Other</LI></UL>]
     */
    constructor(filterCategory: FilterCategory, multiplier: Float?, matrix: FloatArray) : this(filterCategory, Offset(0f), multiplier, matrix) {}

    /**
     * Constructor
     *
     * @param filterCategory Category of effect
     * @param offSet Effect Value offset
     * @param matrix [<UL><LI>4X5 for Color</LI><LI>3X3 for Conventional</LI><LI>1X1 for Rest Other</LI></UL>]
     */
    constructor(filterCategory: FilterCategory, offSet: Offset, matrix: FloatArray) : this(filterCategory, offSet, 1f, matrix) {}

    /**
     * Constructor
     *
     * @param filterCategory Category of effect
     * @param offSet Effect Value offset
     * @param multiplier Effect Value Multiplier
     * @param matrix [<UL><LI>4X5 for Color</LI><LI>3X3 for Conventional</LI><LI>1X1 for Rest Other</LI></UL>]
     */
    constructor(filterCategory: FilterCategory, offSet: Offset, multiplier: Float?, matrix: FloatArray) {
        this.filterCategory = filterCategory
        this.multiplier = multiplier!!
        this.offSet = offSet
        this.matrix = getMatrix(filterCategory, offSet, matrix)
    }

    /**
     * Method to create Filter mMatrix from Input mMatrix
     * @param filterCategory
     * @param offSet
     * @param matrix
     */
    private fun getMatrix(filterCategory: FilterCategory, offSet: Offset, matrix: FloatArray): FloatArray {
        var matrix = matrix
        if (matrix.isEmpty()) {
            when (filterCategory) {
                FilterCategory.COLOR -> matrix = floatArrayOf(1f, 0f, 0f, 0f, 1f, 0f, 0f, 0f, 1f)
                FilterCategory.CONVENTIONAL -> matrix = floatArrayOf(0f, 0f, 0f, 0f, 1f, 0f, 0f, 0f, 10f)
                else -> matrix = floatArrayOf(0f)
            }
        }
        when (filterCategory) {
            FilterCategory.COLOR -> return floatArrayOf(matrix[0], matrix[1], matrix[2], 0f, offSet.r, matrix[3], matrix[4], matrix[5], 0f, offSet.b, matrix[6], matrix[7], matrix[8], 0f, offSet.g, 0f, 0f, 0f, 1f, 0f)
            FilterCategory.CONVENTIONAL -> return floatArrayOf(matrix[0], matrix[1], matrix[2], matrix[3], matrix[4], matrix[5], matrix[6], matrix[7], matrix[8])
            else -> return floatArrayOf(0f)
        }
    }
}